x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -0,0 +1,238 @@
+
+    private final Client clientWithOrigin;
+    private final String systemIndexName;
+    private final ThreadPool threadPool;
+
+    public BlobStoreCacheMaintenanceService(ThreadPool threadPool, Client client, String systemIndexName) {
+        this.clientWithOrigin = new OriginSettingClient(Objects.requireNonNull(client), SEARCHABLE_SNAPSHOTS_ORIGIN);
+        this.systemIndexName = Objects.requireNonNull(systemIndexName);
+        this.threadPool = Objects.requireNonNull(threadPool);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        final ClusterState state = event.state();
+        if (state.getBlocks().hasGlobalBlock(STATE_NOT_RECOVERED_BLOCK)) {
+            return; // state not fully recovered
+        }
+        final ShardRouting primary = systemIndexPrimaryShard(state);
+        if (primary == null || Objects.equals(state.nodes().getLocalNodeId(), primary.currentNodeId()) == false) {
+            return; // system index primary shard does not exist or is not assigned to this data node
+        }
+        if (event.indicesDeleted().isEmpty() == false) {
+            threadPool.generic().execute(new MaintenanceTask(event));
+        }
+    }
+
+    @Nullable
+    private ShardRouting systemIndexPrimaryShard(final ClusterState state) {
+        final IndexMetadata indexMetadata = state.metadata().index(systemIndexName);
+        if (indexMetadata != null) {
+            final IndexRoutingTable indexRoutingTable = state.routingTable().index(indexMetadata.getIndex());
+            if (indexRoutingTable != null) {
+                return indexRoutingTable.shard(0).primaryShard();
+            }
+        }
+        return null;
+    }
+
+    private static boolean hasSearchableSnapshotWith(final ClusterState state, final SnapshotId snapshotId, final IndexId indexId) {
+        for (IndexMetadata indexMetadata : state.metadata()) {
+            final Settings indexSettings = indexMetadata.getSettings();
+            if (SearchableSnapshotsSettings.isSearchableSnapshotStore(indexSettings)) {
+                final SnapshotId otherSnapshotId = new SnapshotId(
+                    SNAPSHOT_SNAPSHOT_NAME_SETTING.get(indexSettings),
+                    SNAPSHOT_SNAPSHOT_ID_SETTING.get(indexSettings)
+                );
+                if (Objects.equals(snapshotId, otherSnapshotId)) {
+                    final IndexId otherIndexId = new IndexId(
+                        SNAPSHOT_INDEX_NAME_SETTING.get(indexSettings),
+                        SNAPSHOT_INDEX_ID_SETTING.get(indexSettings)
+                    );
+                    if (Objects.equals(indexId, otherIndexId)) {
+                        return true;
+                    }
+                }
+            }
+        }
+        return false;
+    }
+
+    static QueryBuilder buildDeleteByQuery(int numberOfShards, String snapshotUuid, String indexUuid) {
+        final Set<String> paths = new HashSet<>(numberOfShards);
```

Comment:
Two comments:

* `HashSet` accepts the capacity, not the amount of expected elements in the constructor, so it would actually be re-sized in the runtime.
* Consider using the Streams API here to avoid manually sizing the set:
```java
IntStream.range(0, numberOfShards)
            .mapToObj(shard -> String.join("/", snapshotUuid, indexUuid, String.valueOf(shard)))
            .collect(Collectors.toSet())
```

Reply:
Thanks for catching this, looks like an ArrayList got changed into a HashSet 😬 I pushed d8164199dde5c00be5670b1b390ce511f0fdf666